### DISCUSSION TITLE
About nebula-spark-connector v3.0.0

### DISCUSSION THREAD

**Discussion by porscheme** [vesoft-inc/nebula#4231] [Q&A]
Nebula: v3.1.0
Spark Connector: 3.0.0
Spark: 3.2.1

1. Our spark and nebula are in two different K8S clusters. Therefore, what how can spark jobs talk to Nebula cluster using nebula-spark-connector?
2. Has anyone successfully used nebula-spark-connector (written in Scala?) in pySpark? If, example will be very help full.

**Comment by wey-gu** [comment]
1. To connect to nebula from a spark running in a different cluster, we need to expose all meta/storage out of the cluster, and explicitly set them to the reader rather than only specifying the meta(then it will fetch storage from meta only), ~~I will look into this later(now on my phone replying you)~~ now I remember that we need this to be implemented to enable it https://github.com/vesoft-inc/nebula-spark-connector/issues/39
2. Scala/Pyspark runnable examples: https://www.siwei.io/en/spark-on-nebula-graph/

**Comment by porscheme** [comment]
- @wey-gu we used [Helm Charts](https://github.com/vesoft-inc/nebula-operator/tree/master/charts) to deploy Nebula
- Unfortunately, it does have any method to expose metad/storaged out of the cluster
- AKS has something called internal Loadbalancer (not exposed to the internet). We use this feature to expose Elasticsearch to the applications, for example

Can Nebula engineering team add ability to expose metad/storaged service outside the cluster?

**Comment by porscheme** [comment]
Even if I make were to follow @MegaByte875 suggestion to expose graphd  example here https://github.com/vesoft-inc/nebula-operator/tree/master/config/samples, how can expose meta? 

Looks like gap in the solution, confirm?

We are blocked!

Any alternate solution is really appreciated.

**Reply by wey-gu** [reply]
expose them should be common experiences in k8s, I could provide poc for that, before that we need spark-connector to support specifying storaged addresses instead of fetching from meta only, which need changes in nebula-Java and nebula-spark-connector. And I had been working on pr of this since this morning for you:).

Hopefully I could make them done shortly, I got you back, no worries 😆

**Comment by porscheme** [comment]
Awesome, thanks looking forward to your message.

**Comment by wey-gu** [comment]
**progress**

- [ ] Nebula-Java support specifying storageD addresses https://github.com/vesoft-inc/nebula-java/pull/451
- [ ] Nebula-Spark-Connecotr support specifying storageD addresses https://github.com/vesoft-inc/nebula-spark-connector/pull/47
- [ ] PoC of exposing meta, storage instance addresses

**Comment by wey-gu** [comment]
Update, after my PR on nebula-java and nebula-spark-c to support specifying exposed storaged addresses being discussed in the review, it turned out this is not easily feasible due to the map of storage-hosts to their address was used when seeking data distribution among storage hosts, thus simply specify list of storage hosts isn't enough, their map should be configurable, too, this design is more complex, thus need further discussion.

Thus, I created a PoC to only expose them and use TCP proxy and DNS to make spark-connector connect to nebula with exactly the same addresses as it was inside the nebula cluster namespace:

a. create services per metad and storaged pod with LoadBalancer type to expose outside the cluster b. Use TCP Proxy(b.1) or/and DNS(b.2) to resolve the endpoint as their: inside the cluster

In this example, in b. I give a demo where both TCP Proxy and DNS are used, actually, only DNS is enough if the exposed port remains the same of src and target.

details refer to: https://gist.github.com/wey-gu/699b9a2ef5dff5f0fb5f288d692ddfd5

**Reply by sunkararp** [reply]
> Update, after my PR on nebula-java and nebula-spark-c to support specifying exposed storaged addresses being discussed in the review, it turned out this is not easily feasible due to the map of storage-hosts to their address was used when seeking data distribution among storage hosts, thus simply specify list of storage hosts isn't enough, their map should be configurable, too, this design is more complex, thus need further discussion.
> 
> Thus, I created a PoC to only expose them and use TCP proxy and DNS to make spark-connector connect to nebula with exactly the same addresses as it was inside the nebula cluster namespace:
> 
> a. create services per metad and storaged pod with LoadBalancer type to expose outside the cluster b. Use TCP Proxy(b.1) or/and DNS(b.2) to resolve the endpoint as their: inside the cluster
> 
> In this example, in b. I give a demo where both TCP Proxy and DNS are used, actually, only DNS is enough if the exposed port remains the same of src and target.
> 
> details refer to: https://gist.github.com/wey-gu/699b9a2ef5dff5f0fb5f288d692ddfd5

Any update on this work item?

**Comment by porscheme** [comment]
- I created [graphd-service.yaml](https://github.com/vesoft-inc/nebula/files/8666363/graphd-service.yaml.txt) and [metad-service.yaml](https://github.com/vesoft-inc/nebula/files/8666364/metad-service.yaml.txt) YAML files
- Placed above two YAML in /nebula-cluster/templates folder (see attached) ![nebula-cluster-tree](https://user-images.githubusercontent.com/29828396/167760130-670cca4e-56a8-4830-9523-a6ff33587e20.jpg)
- Using [this](https://github.com/vesoft-inc/nebula/files/8666386/test.nebula.values.yaml.txt) helm charts values file I was able to register internal loadbalance for both graphd and metad.
- I was able to connect and query using nebula3-python python package from Jupyter notebook, I have not tried spark-connector yet
- Please review attached files and provide any feedback, if nay

**Reply by wey-gu** [reply]
I don't think this would work, in GraphD is stateless, thus exposing it with LB should be fine. While Meta/Storage are stateful instances for meta client and storage client, putting them behind one load balancer round-robin-ly is not supported. For spark-connector, the normal pattern was to(not like every other clients) scan data via storage client, it requires to get the data distribution from meta and then scanning storage instances based on that distribution.

**Comment by porscheme** [comment]
@wey-gu thanks for the reply.

- I did not change Storage service
- Even though meta is exposed as LB there is only replica. Do we need more than one meta replica? 
- Also, load balancing doesn't need be round-robin. check [this](https://docs.microsoft.com/en-us/azure/load-balancer/distribution-mode-concepts)

**Reply by wey-gu** [reply]
> - I did not change Storage service

Not like other clients, the spark connector of Nebula Graph was designed for analytical purposes, thus it's fetching data with StorageClient (instead of querying with nGQL through GraphD), the storage client was designed to be aware of each of the storage instances, this means all storage instances' addresses should be exposed separately.

> - Even though meta is exposed as LB there is only replica. Do we need more than one meta replica?

If Meta replication is set 1, the LB model would work perfectly, in this case, it's recommended to have its data replicated with PV.

> - Also, load balancing doesn't need be round-robin. check [this](https://docs.microsoft.com/en-us/azure/load-balancer/distribution-mode-concepts)

Sorry, I had put it wrongly, I didn't mean it's the problem of the round-robin(which should be applied to all LB mechanisms). As spark client was designed to scan data from storaged, it needs to know each instance of them instead of a load-balanced single point.
